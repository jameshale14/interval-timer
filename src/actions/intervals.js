import database from '../firebase/firebase'

//GET_INTERVALS
export const getIntervals = () => ({
  type: 'GET_INTERVALS',
})

//START_GET_INTERVALS

//CREATE_INTERVAL
export const createInterval = (interval) => ({
  type: 'CREATE_INTERVAL',
  interval
})

//START_CREATE_INTERVAL

//UPDATE_INTERVAL
export const updateInterval = (id, updates) => ({
  type: 'UPDATE_INTERVAL',
  id,
  updates
})

//START_UPDATE_INTERVAL

//DELETE_INTERVAL
export const deleteInterval = (id) => ({
  type: 'DELETE_INTERVAL',
  id
})

//START_DELETE_INTERVAL

export const startDeleteInterval = (id) => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid
    return database.ref(`users/${uid}/intervals/${id}`).remove()
      .then(() => {
        dispatch(deleteInterval(id))
      })
  }
}

//SET_INTERVALS
export const setIntervals = (intervals) => ({
  type: 'SET_INTERVALS',
  intervals
})

//START_SET_INTERVALS
export const startSetIntervals = () => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid
    return database.ref(`users/${uid}/intervals`).once('value')
      .then((snapshot) => {
        const intervals = []
        snapshot.forEach((childSnapshot) => {
          intervals.push({
            id: childSnapshot.key,
            ...childSnapshot.val()
          })
        })
        dispatch(setIntervals(intervals))
      })
  }
}