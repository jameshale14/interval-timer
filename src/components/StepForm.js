import React from 'react'
import PropTypes from 'prop-types'

export default class StepForm extends React.Component {

  constructor(props) {
    super(props)
    this.state = {
      error: undefined,
      isRestType: false,
      type: 'Activity',
      name: '',
      duration: 0
    }
  }

  onTypeChange = (e) => {
    const type = e.target.value
    const isRestType = type == 'Rest' ? true : false
    const name = isRestType ? 'Rest' : ''
    this.setState(() => ({ type, isRestType, name }))
  }

  onNameChange = (e) => {
    const name = e.target.value
    this.setState(() => ({ name }))
  }

  onDurationChange = (e) => {
    const duration = e.target.value
    this.setState(() => ({ duration }))
  }

  handleAddStep = (e) => {
    e.preventDefault()
    if (this.state.type == '' || this.state.name == '' || this.state.duration == 0) {
      this.setState(() => ({ error: 'Please fill in all fields' }))
    } else {
      this.setState(() => ({ error: undefined }))
      const newStep = {
        name: this.state.name,
        type: this.state.type,
        duration: this.state.duration
      }
      this.props.onSubmit(newStep)
    }
  }

  render() {
    return (
      <div>
        {this.state.error && <p>{this.state.error}</p>}
        <form onSubmit={this.handleAddStep}>
          <select
            value={this.state.type}
            onChange={this.onTypeChange}
            onBlur={this.onTypeChange}
          >
            <option value='Activity'>Activity</option>
            <option value='Rest'>Rest</option>
          </select>

          <input
            type='text'
            placeholder='Step name'
            value={this.state.name}
            onChange={this.onNameChange}
            disabled={this.state.isRestType}
          />
          <input
            type='number'
            placeholder='Duration (seconds)'
            step='1'
            value={this.state.duration}
            onChange={this.onDurationChange}
          />
          <button onClick={this.handleAddStep}>Add Step</button>
        </form>
      </div>
    )
  }
}

StepForm.propTypes = {
  onSubmit: PropTypes.func
}