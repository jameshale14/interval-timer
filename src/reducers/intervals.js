export default (state = {}, action) => {
  switch (action.type) {
    case 'GET_INTERVALS':
      return state
    case 'CREATE_INTERVAL':
      return [...state, action.interval]
    case 'UPDATE_INTERVAL':
      return state.map((interval) => {
        if (interval.id === action.id) {
          return {
            ...interval,
            ...action.updates
          }
        } else {
          return interval
        }
      })
    case 'DELETE_INTERVAL':
      return state.filter((interval) => {
        return interval.id !== action.id
      })
    case 'SET_INTERVALS':
      return action.intervals
    default:
      return state
  }
}