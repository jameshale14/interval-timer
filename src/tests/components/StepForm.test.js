import React from 'React'
import { shallow } from 'enzyme'
import StepForm from '../../components/StepForm'
import { intervals } from '../fixtures/intervals'

let wrapper, onSubmit

beforeEach(() => {
  onSubmit = jest.fn()
  wrapper = shallow(<StepForm onSubmit={onSubmit} />)
})

test('should render StepForm correctly', () => {
  expect(wrapper).toMatchSnapshot()
})

test('should throw error when adding step with no data', () => {
  wrapper.find('form').simulate('submit', {
    preventDefault: () => { }
  })
  expect(wrapper.state().error.length).toBeGreaterThan(0)
  expect(wrapper).toMatchSnapshot()
})

test('should call onSubmit when all data is provided', () => {

  //set the type of step
  wrapper.find('select').at(0).simulate('change', {
    target: { value: 'Activity' }
  })
  expect(wrapper.state().type).toEqual('Activity')
  expect(wrapper).toMatchSnapshot()

  //set the step name
  wrapper.find('input').at(0).simulate('change', {
    target: { value: 'Pull-ups' }
  })
  expect(wrapper.state().name).toEqual('Pull-ups')
  expect(wrapper).toMatchSnapshot()

  //set the step duration
  wrapper.find('input').at(1).simulate('change', {
    target: { value: 20 }
  })
  expect(wrapper.state().duration).toEqual(20)
  expect(wrapper).toMatchSnapshot()

  //add the step to the interval routine
  wrapper.find('button').at(0).simulate('click', {
    preventDefault: () => { }
  })

  expect(wrapper.state().error).toBe(undefined)
  expect(wrapper).toMatchSnapshot()
})

test('should disable the input for step name when step type is set to Rest', () => {
  wrapper.find('select').at(0).simulate('change', {
    target: { value: 'Rest' }
  })
  expect(wrapper.state().isRestType).toBe(true)
  expect(wrapper.state().name).toEqual('Rest')
  expect(wrapper).toMatchSnapshot()

  // Try to change the input, make sure it is disabled
  expect(wrapper.find('input').at(0).prop('disabled')).toBe(true)

})

test('should re-enable the input for step name when step type is set back to Activity', () => {
  wrapper.find('select').at(0).simulate('change', {
    target: { value: 'Rest' }
  })

  wrapper.find('select').at(0).simulate('change', {
    target: { value: 'Activity' }
  })

  // Check the name input is NOT disabled
  expect(wrapper.find('input').at(0).prop('disabled')).toBe(false)
  expect(wrapper.state().name).toEqual('')
}) 