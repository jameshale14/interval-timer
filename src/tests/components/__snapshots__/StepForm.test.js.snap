// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should call onSubmit when all data is provided 1`] = `
<div>
  <form
    onSubmit={[Function]}
  >
    <select
      onBlur={[Function]}
      onChange={[Function]}
      value="Activity"
    >
      <option
        value="Activity"
      >
        Activity
      </option>
      <option
        value="Rest"
      >
        Rest
      </option>
    </select>
    <input
      disabled={false}
      onChange={[Function]}
      placeholder="Step name"
      type="text"
      value=""
    />
    <input
      onChange={[Function]}
      placeholder="Duration (seconds)"
      step="1"
      type="number"
      value={0}
    />
    <button
      onClick={[Function]}
    >
      Add Step
    </button>
  </form>
</div>
`;

exports[`should call onSubmit when all data is provided 2`] = `
<div>
  <form
    onSubmit={[Function]}
  >
    <select
      onBlur={[Function]}
      onChange={[Function]}
      value="Activity"
    >
      <option
        value="Activity"
      >
        Activity
      </option>
      <option
        value="Rest"
      >
        Rest
      </option>
    </select>
    <input
      disabled={false}
      onChange={[Function]}
      placeholder="Step name"
      type="text"
      value="Pull-ups"
    />
    <input
      onChange={[Function]}
      placeholder="Duration (seconds)"
      step="1"
      type="number"
      value={0}
    />
    <button
      onClick={[Function]}
    >
      Add Step
    </button>
  </form>
</div>
`;

exports[`should call onSubmit when all data is provided 3`] = `
<div>
  <form
    onSubmit={[Function]}
  >
    <select
      onBlur={[Function]}
      onChange={[Function]}
      value="Activity"
    >
      <option
        value="Activity"
      >
        Activity
      </option>
      <option
        value="Rest"
      >
        Rest
      </option>
    </select>
    <input
      disabled={false}
      onChange={[Function]}
      placeholder="Step name"
      type="text"
      value="Pull-ups"
    />
    <input
      onChange={[Function]}
      placeholder="Duration (seconds)"
      step="1"
      type="number"
      value={20}
    />
    <button
      onClick={[Function]}
    >
      Add Step
    </button>
  </form>
</div>
`;

exports[`should call onSubmit when all data is provided 4`] = `
<div>
  <form
    onSubmit={[Function]}
  >
    <select
      onBlur={[Function]}
      onChange={[Function]}
      value="Activity"
    >
      <option
        value="Activity"
      >
        Activity
      </option>
      <option
        value="Rest"
      >
        Rest
      </option>
    </select>
    <input
      disabled={false}
      onChange={[Function]}
      placeholder="Step name"
      type="text"
      value="Pull-ups"
    />
    <input
      onChange={[Function]}
      placeholder="Duration (seconds)"
      step="1"
      type="number"
      value={20}
    />
    <button
      onClick={[Function]}
    >
      Add Step
    </button>
  </form>
</div>
`;

exports[`should disable the input for step name when step type is set to Rest 1`] = `
<div>
  <form
    onSubmit={[Function]}
  >
    <select
      onBlur={[Function]}
      onChange={[Function]}
      value="Rest"
    >
      <option
        value="Activity"
      >
        Activity
      </option>
      <option
        value="Rest"
      >
        Rest
      </option>
    </select>
    <input
      disabled={true}
      onChange={[Function]}
      placeholder="Step name"
      type="text"
      value="Rest"
    />
    <input
      onChange={[Function]}
      placeholder="Duration (seconds)"
      step="1"
      type="number"
      value={0}
    />
    <button
      onClick={[Function]}
    >
      Add Step
    </button>
  </form>
</div>
`;

exports[`should render StepForm correctly 1`] = `
<div>
  <form
    onSubmit={[Function]}
  >
    <select
      onBlur={[Function]}
      onChange={[Function]}
      value="Activity"
    >
      <option
        value="Activity"
      >
        Activity
      </option>
      <option
        value="Rest"
      >
        Rest
      </option>
    </select>
    <input
      disabled={false}
      onChange={[Function]}
      placeholder="Step name"
      type="text"
      value=""
    />
    <input
      onChange={[Function]}
      placeholder="Duration (seconds)"
      step="1"
      type="number"
      value={0}
    />
    <button
      onClick={[Function]}
    >
      Add Step
    </button>
  </form>
</div>
`;

exports[`should throw error when adding step with no data 1`] = `
<div>
  <p>
    Please fill in all fields
  </p>
  <form
    onSubmit={[Function]}
  >
    <select
      onBlur={[Function]}
      onChange={[Function]}
      value="Activity"
    >
      <option
        value="Activity"
      >
        Activity
      </option>
      <option
        value="Rest"
      >
        Rest
      </option>
    </select>
    <input
      disabled={false}
      onChange={[Function]}
      placeholder="Step name"
      type="text"
      value=""
    />
    <input
      onChange={[Function]}
      placeholder="Duration (seconds)"
      step="1"
      type="number"
      value={0}
    />
    <button
      onClick={[Function]}
    >
      Add Step
    </button>
  </form>
</div>
`;
